File: packages/point/python_std/point_std/crypto.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
from typing import Any


def _base64url_encode(value: str) -> str:
	return base64.urlsafe_b64encode(value.encode("utf-8")).decode("ascii").rstrip("=")


def _base64url_decode(value: str) -> str:
	padding = "=" * ((4 - len(value) % 4) % 4)
	return base64.urlsafe_b64decode(value + padding).decode("utf-8")


def _normalize_bearer_token(token: str | None) -> str:
	if token is None:
		return ""
	trimmed = token.strip()
	if trimmed.startswith("Bearer "):
		return trimmed[len("Bearer ") :].strip()
	return trimmed


def cryptoSha256(value: str) -> str:
	return hashlib.sha256(value.encode("utf-8")).hexdigest()


def cryptoHmacSha256(value: str, secret: str) -> str:
	return hmac.new(secret.encode("utf-8"), value.encode("utf-8"), hashlib.sha256).hexdigest()


def cryptoJwtSign(payload: str, secret: str) -> str:
	json.loads(payload)
	header = _base64url_encode(json.dumps({"alg": "HS256", "typ": "JWT"}, separators=(",", ":")))
	body = _base64url_encode(payload)
	data = f"{header}.{body}"
	signature = hmac.new(secret.encode("utf-8"), data.encode("utf-8"), hashlib.sha256).digest()
	signature_text = base64.urlsafe_b64encode(signature).decode("ascii").rstrip("=")
	return f"{data}.{signature_text}"


def cryptoJwtVerify(token: str | None, secret: str) -> str | dict[str, str]:
	normalized = _normalize_bearer_token(token)
	if not normalized:
		return {"message": "Missing JWT"}
	parts = normalized.split(".")
	if len(parts) != 3:
		return {"message": "Invalid JWT"}
	header, body, signature = parts
	data = f"{header}.{body}"
	expected = hmac.new(secret.encode("utf-8"), data.encode("utf-8"), hashlib.sha256).digest()
	expected_text = base64.urlsafe_b64encode(expected).decode("ascii").rstrip("=")
	if len(expected_text) != len(signature) or not hmac.compare_digest(expected_text, signature):
		return {"message": "Invalid JWT signature"}
	try:
		payload = _base64url_decode(body)
		json.loads(payload)
		return payload
	except Exception as error:
		return {"message": str(error)}


def cryptoJwtIsValid(token: str | None, secret: str) -> bool:
	return isinstance(cryptoJwtVerify(token, secret), str)


sha256Hash = cryptoSha256
hmacSha256 = cryptoHmacSha256
jwtSign = cryptoJwtSign
jwtVerify = cryptoJwtVerify
checkJwtValid = cryptoJwtIsValid

File: packages/point/src/std/crypto.ts
import { createHash, createHmac, timingSafeEqual } from "node:crypto";

export type PointStdError = { message: string };

function base64UrlEncode(value: string): string {
	return Buffer.from(value, "utf8").toString("base64url");
}

function base64UrlDecode(value: string): string {
	return Buffer.from(value, "base64url").toString("utf8");
}

function normalizeBearerToken(token: string | null | undefined): string {
	if (token == null) return "";
	const trimmed = token.trim();
	return trimmed.startsWith("Bearer ") ? trimmed.slice("Bearer ".length).trim() : trimmed;
}

export function cryptoSha256(value: string): string {
	return createHash("sha256").update(value, "utf8").digest("hex");
}

export function cryptoHmacSha256(value: string, secret: string): string {
	return createHmac("sha256", secret).update(value, "utf8").digest("hex");
}

export function cryptoJwtSign(payload: string, secret: string): string {
	JSON.parse(payload);
	const header = base64UrlEncode(JSON.stringify({ alg: "HS256", typ: "JWT" }));
	const body = base64UrlEncode(payload);
	const data = `${header}.${body}`;
	const signature = createHmac("sha256", secret).update(data, "utf8").digest("base64url");
	return `${data}.${signature}`;
}

export function cryptoJwtVerify(token: string | null | undefined, secret: string): string | PointStdError {
	const normalized = normalizeBearerToken(token);
	if (!normalized) {
		return { message: "Missing JWT" };
	}
	const parts = normalized.split(".");
	if (parts.length !== 3) {
		return { message: "Invalid JWT" };
	}
	const [header, body, signature] = parts;
	const data = `${header}.${body}`;
	const expected = createHmac("sha256", secret).update(data, "utf8").digest("base64url");
	try {
		const expectedBuffer = Buffer.from(expected, "utf8");
		const signatureBuffer = Buffer.from(signature, "utf8");
		if (expectedBuffer.length !== signatureBuffer.length || !timingSafeEqual(expectedBuffer, signatureBuffer)) {
			return { message: "Invalid JWT signature" };
		}
		const payload = base64UrlDecode(body);
		JSON.parse(payload);
		return payload;
	} catch (error) {
		return { message: error instanceof Error ? error.message : String(error) };
	}
}

export function cryptoJwtIsValid(token: string | null | undefined, secret: string): boolean {
	return typeof cryptoJwtVerify(token, secret) === "string";
}

export {
	cryptoSha256 as sha256Hash,
	cryptoHmacSha256 as hmacSha256,
	cryptoJwtSign as jwtSign,
	cryptoJwtVerify as jwtVerify,
	cryptoJwtIsValid as checkJwtValid,
};
